Separating Frontend and Backend Functionality in NDT HTML5 Client - Part 1

shredtechular opened the pull request:
Per feedback contained in #49 this PR splits the changes in that PR into a more manageable PR.  This first PR contains and only contains the changes needed to separate the backend functionality from the `script.js` file into an `ndt-library.js` file.  When moving that functionality to the `ndt-library.js` file, it needs to be removed from the `script.js` file and then minimal changes to make sure the application still works.

The idea is to ease how an implementer might include an NDT js library onto their website and by including only the necessary js, an implementer could create their own frontend using the NDT js library.  

As the current implementation would require a website to include [five different scripts](https://github.com/m-lab/m-lab.github.io/pull/76/files#diff-56ff23ff6448049981daee3c0344f9cdR22), this PR starts to address that breaking up the [script.js](https://github.com/m-lab/ndt/blob/master/HTML5-frontend/script.js) file into frontend and backend functionality.  This removes the dependency for an NDT js library to need jQuery, the gauges.min.js library, or the script.js file.  Those files would **ONLY** be needed if a user wanted the NDT frontend that exists on the site today, but not if it was creating a custom frontend.

1 small change was made to the css and `ndt-browser-client.js`  files in order to make the application still function properly.

Finally, also added the ndt-library.js include into the widget.html.

Once this PR and part 2 #55 is merged, #49 can be closed.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/ndt/54)
<!-- Reviewable:end -->

critzo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> But I would also like a second reviewer to take a look. @fernandalavalle or @mtlynch ?

<details><summary><i>Previously, shredtechular (Andrew Newhouse) wrote…</i></summary>
> #### Separating Frontend and Backend Functionality in NDT HTML5 Client \- Part 1
> 
> Per feedback contained in #49 this PR splits the changes in that PR into a more manageable PR.  This first PR contains and only contains the changes needed to separate the backend functionality from the `script.js` file into an `ndt-library.js` file.  When moving that functionality to the `ndt-library.js` file, it needs to be removed from the `script.js` file and then minimal changes to make sure the application still works.
> 
> 
> 
> The idea is to ease how an implementer might include an NDT js library onto their website and by including only the necessary js, an implementer could create their own frontend using the NDT js library.  
> 
> 
> 
> As the current implementation would require a website to include [five different scripts](https://github.com/m-lab/m-lab.github.io/pull/76/files#diff-56ff23ff6448049981daee3c0344f9cdR22), this PR starts to address that breaking up the [script.js](https://github.com/m-lab/ndt/blob/master/HTML5-frontend/script.js) file into frontend and backend functionality.  This removes the dependency for an NDT js library to need jQuery, the gauges.min.js library, or the script.js file.  Those files would **ONLY** be needed if a user wanted the NDT frontend that exists on the site today, but not if it was creating a custom frontend.
> 
> 
> 
> 1 small change was made to the css and `ndt-browser-client.js`  files in order to make the application still function properly.
> 
> 
> 
> Finally, also added the ndt-library.js include into the widget.html.
> 
> 
> 
> Once this PR and part 2 #55 is merged, #49 can be closed.
> 
> 
> 
> 
> 
> </details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/54#-:-KHtWRBo60U-TVm8P44h:-195709907)*
<!-- Sent from Reviewable.io -->

collina commented on HTML5-frontend/ndt-browser-client.js:
```diff
@@ -84,6 +84,7 @@ NDTjs.prototype.logger = function (logMessage, debugging) {
  * @returns {boolean} Browser supports necessary functions for test client.
  */
 NDTjs.prototype.checkBrowserSupport = function () {
+  if (typeof window != "undefined") self = window;
```
Does our style guideline allow for one line conditionals?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+/**
+ * @fileoverview This javascript code includes an NDT object that contains the
```
I think it's actually the reverse – the other file is the library and this is the interface that provides callbacks and m-lab-ns support.

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+  'use strict';
+
+  /** @constructor */
+  function NDTLibjs() {
```
NDTInterface?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+   * websockets.  If not, will attempt to use Java if available.
+   */
+  NDTLibjs.prototype.checkInstalledPlugins = function() {
+    var hasJava = false;
```
Aren't we dropping Java?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+
+    if (hasWebsockets) {
+      this.use_websocket_client = true;
+
```
Empty space.

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    if (hasWebsockets) {
+      this.use_websocket_client = true;
```
First, I suspect we would consider this a private class property, but why do we set it that way, instead of returning it?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+        hasJava = false;
+      }
+    }
+    hasWebsockets = false;
```
already false in instantiation?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+  /**
+   * Determines where to run the tests.  If against the websocket or Java
+   * applet
+   * @return - {obj} websocket client or java applet element id
```
"Object"?

collina replied:
HTML Entity?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+   * @return - none, sets object property for server
+   */
+  NDTLibjs.prototype.getServer = function () {
+    var mlabNsService = 'https:' == location.protocol ? 'ndt_ssl' : 'ndt';
```
Style complaint? Either way, please break out it's a bit obtuse.

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+    var mlabNsUrl = 'https://mlab-ns.appspot.com/';
+    self = this;
+
+    var NDTAjax = new XMLHttpRequest();
```
Declarations at start of function.

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+    NDTAjax.onreadystatechange = function () {
+      // completed and successful
+      if (NDTAjax.readyState == 4 && NDTAjax.status == 200) {
+        var resp = JSON.parse(NDTAjax.responseText);
```
Declarations at start of function.

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+  NDTLibjs.prototype.getServer = function () {
+    var mlabNsService = 'https:' == location.protocol ? 'ndt_ssl' : 'ndt';
+    var mlabNsUrl = 'https://mlab-ns.appspot.com/';
+    self = this;
```
We need to standardize on "self" or "that" for dealing with scope issues.

collina commented on HTML5-frontend/script.js:
```diff
@@ -8,7 +8,6 @@ $(function(){
     setTimeout(initializeTest, 1000);
```
Can we at this point change the script ("script.js") filename to something that is more understandable?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+
+  // if not already created, instantiate object just in case
+  if (typeof(NDT) === 'undefined') {
+    window.NDT = new NDTLibjs();
```
Let's give it a bit more specific of a instance name (NDTObject or something)?

collina replied:
As mentioned elsewhere, can this be done outside of a property of window?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+      return true;
+    }
+    else {
+      var app = document.createElement('applet');
```
var declaration beyond start of function.

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+   */
+  NDTLibjs.prototype.uploadSpeed = function(raw) {
+    if (this.simulate) return 0;
+    var speed = this.testNDT().getNDTvar('ClientToServerSpeed');
```
var declaration beyond start of function.

collina commented on HTML5-frontend/script.js:
```diff
@@ -65,19 +58,23 @@ function initializeTest() {
+    return simulateTest();
+  } else {
+    $('#backendContainer').empty();
+    var clientProtocol = NDT.startTest();
```
var declaration beyond start of function.

collina commented on HTML5-frontend/script.js:
```diff
@@ -499,7 +449,7 @@ function testDetails() {
   var d = '';
 
-  var errorMsg = testError();
+  var errorMsg = NDT.testError();
```
var declaration beyond start of function.

pboothe commented:
I am taking a look right now.

<details><summary><i>Previously, critzo wrote…</i></summary>
> :lgtm: But I would also like a second reviewer to take a look. @fernandalavalle or @mtlynch ?</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/54#-:-KHvAVKGN4VXk1eeTlgI:619003802)*
<!-- Sent from Reviewable.io -->

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+
+    // Initializes simulate property to false.  Can be changed upon starting a
+    // test in the testServer method
+    this.simulate = false;
```
Since these are private instance variables do we denote them in any decorator or a prepended underscore?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+
+  /**
+   * Operates on an instance of NDT to begin executing a test.
+   * @param - none
```
Is this right?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+  /**
+   * Operates on an instance of NDT to begin executing a test.
+   * @param - none
+   * @return - {string} True if client is using websockets, False if error
```
bool?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+   * Operates on an instance of NDT to begin executing a test.
+   * @param - none
+   * @return - {string} True if client is using websockets, False if error
+   * occurred during finding backend server, or applet container element fall
```
This seems overloaded – a bool or an applet container?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+   * back to Java applet.
+   */
+  NDTLibjs.prototype.startTest = function() {
+    if (NDT.ndtServer) {
```
I'm not sure that the class should expect NDT to be a window properties rather than an object passed in.

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+  NDTLibjs.prototype.startTest = function() {
+    if (NDT.ndtServer) {
+      this.checkInstalledPlugins();
+      var clientProtocol = this.createBackend();
```
var declaraction

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+      this.testNDT().run_test(this.ndtServer);
+    } else {
+      // in case AJAX call fails or is taking longer
+      this.getServer();
```
I don't understand, if getServer fails, it does getServer again and then returns false?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+    // Look up a suitable NDT server using mlab-ns. The ndtServer property is
+    //  used by the backend javascript to determine which server to test
+    //  against.
+    this.getServer();
```
This is asynchronous, right? If so, this won't block and the rest might not be populated as it progresses.

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+   *     otherwise returns false
+   */
+  NDTLibjs.prototype.createBackend = function() {
+    if (this.use_websocket_client) {
```
I would opt for these functions to be populated with variables like use_websocket_client passed as parameters, rather that use instance properties

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+  /**
+   * Creates the backend for the NDT object.  Tries to create websocket backend
+   * if supported, falls back to java applet if websocket not available
+   * @return - {obj} If Java is to be used returns Java applet information,
```
Object

collina replied:
It's more of an HTML entity and not information, right?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+   * Creates the backend for the NDT object.  Tries to create websocket backend
+   * if supported, falls back to java applet if websocket not available
+   * @return - {obj} If Java is to be used returns Java applet information,
+   *     otherwise returns false
```
Where does "false" come from – looks like True or the HTML entity?

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+    var hasJava = false;
+    var hasWebsockets = false;
+
+    hasJava = true;
```
sets to false, then true, then false, I would assume false and then find if true.

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+        hasWebsockets = true;
+      }
+    } catch(e) {
+      hasWebsockets = false;
```
if you assume true, then you might not need this line

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+    }
+    else if (hasJava) {
+      this.use_websocket_client = false;
+
```
empty space

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+      this.use_websocket_client = false;
+
+    }
+  }
```
if neither hasWebsocket or hasJava?

pboothe commented:
Overall comment so far is that this would benefit from a run through JSLint, it uses globals when it doesn't have to, and we should postpend all private variables with an underscore.  I'll keep going through it (half done) but I wanted to give initial comments right away.

<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> I am taking a look right now.</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/54#-:-KHvF7NoQd17WBFMLHS4:831927234)*
<!-- Sent from Reviewable.io -->

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+
+  /**
+   * Gets the test status of the current NDT test
+   * @return - {sting} status of current test
```
string

collina commented on HTML5-frontend/ndt-library.js:
```diff
@@ -0,0 +1,234 @@
+
+  /**
+   * Gets error messages of the current NDT test
+   * @return - {sting} error message if occurred
```
string

pboothe commented:
Please put 
`/*jslint bitwise: true, browser: true */` 
after the introductory comment in every file (if it's not already there) then run the JSlint commandline tool (or use the jslint website, but the commandline too is easier to run multiple times).

<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> Overall comment so far is that this would benefit from a run through JSLint, it uses globals when it doesn't have to, and we should postpend all private variables with an underscore.  I'll keep going through it (half done) but I wanted to give initial comments right away.</details>


---

Review status: 0 of 5 files reviewed at latest revision, 35 unresolved discussions, all commit checks successful.

---

*[HTML5-frontend/ndt-browser-client.js, line 87 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8g-r1-87:-KHvC819hEO5TEPtrg5V:562451297) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-browser-client.js#L87)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> Does our style guideline allow for one line conditionals?</details>
"Short conditional statements may be written on one line if this enhances readability. You may use this only when the line is brief and the statement does not use the else clause." - from https://google.github.io/styleguide/cppguide.html#Conditionals which is included by reference from https://google.github.io/styleguide/javascriptguide.xml?showone=Code_formatting#Code_formatting

So yes. This is fine.

---

*[HTML5-frontend/ndt-library.js, line 2 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-2:-KHvCJF8BVGa9smDFOW3:630202021) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L2)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> I think it's the reverse – the other file is the library and this is the interface that provides callbacks and m-lab-ns support.</details>
I'd like to call a moratorium on the word Interface here, where we could be using it in both the sense of GUI and in the independent sense of "the abstract object and function signatures which are fleshed out somewhere else".

---

*[HTML5-frontend/ndt-library.js, line 28 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-28:-KHvCnVMKicb4E7txR-1:439129610) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L28)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> Since these are private instance variables do we denote them in any decorator or a prepended underscore? </details>
"Private properties and methods should be named with a trailing underscore.
Protected properties and methods should be named without a trailing underscore (like public ones)."

https://google.github.io/styleguide/javascriptguide.xml?showone=Code_formatting#Naming

Also, "We recommend the use of the JSDoc annotations @private and @protected to indicate visibility levels for classes, functions, and properties." 

https://google.github.io/styleguide/javascriptguide.xml?showone=Code_formatting#Visibility__private_and_protected_fields_

---

*[HTML5-frontend/ndt-library.js, line 38 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-38:-KHvDg8QYJYDg2DWpaI5:-253194001) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L38)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> Is this right?</details>
Agree, please just delete the line.

---

*[HTML5-frontend/ndt-library.js, line 40 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-40:-KHvE51YwbaMzZzyAVTc:-770759467) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L40)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> This seems overloaded – a bool of True/False or an applet container?</details>
A better pattern to use here is to return the test object or to throw an exception if the browser doesn't support websocket NDT.

---

*[HTML5-frontend/ndt-library.js, line 44 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-44:-KHvEOcbar_2iOAHm8rj:1514539327) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L44)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> I'm not sure that the class should expect NDT to be a window properties rather than an object passed in.</details>
I agree here. Please don't use global variables unless you absolutely must.

---

*[HTML5-frontend/ndt-library.js, line 50 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-50:-KHvEg9H9TvVqA1m7nU0:721054780) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L50)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> I don't understand, if getServer fails, it does getServer again and then returns false?</details>
Hopefully refactoring this to use exceptions will make it clearer.

---

*[HTML5-frontend/ndt-library.js, line 81 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-81:-KHvEoLBHtHJU-iIQ053:-335382677) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L81)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> First, I suspect we would consider this a private class property, but why do we set it that way, instead of returning it? i.e. this is "check" but it starts to make decisions. Also why are we falling back? In the interface this is an explicit option that is chosen.
> 
> </details>
I think that offering up a Java Applet is confusing in the year 2016, and we should not do it unless we must.

---

*[HTML5-frontend/ndt-library.js, line 122 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHvFpbitLTcvLILkrBF:-KHvFpbitLTcvLILkrBG:-794934989) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L122)):*
> ```JavaScript
>     if (this.websocket_client) {
>       return this.websocket_client;
>     }
> ```

This is why we have `else {`. This is not a guard, instead the function is 
```javascript
if (x) {
  a();
} else {
  b();
}```

When I describe this function in english, I find myself naturally using the word "otherwise", which means it needs an `else`.

---

*[HTML5-frontend/ndt-library.js, line 132 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-132:-KHvGOmnLNkODwJMfjKO:-1719433933) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L132)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> Style complaint? Either way, please break out it's a bit obtuse.</details>
Reads better with parens (because nobody remembers operator precedence rules outside of PEMDAS).
```javascript
var mlabNsService = ('https:' == location.protocol) ? 'ndt_ssl' : 'ndt';
```

---

*[HTML5-frontend/ndt-library.js, line 134 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-134:-KHvGjA6EF3GqO95Agxz:-2112308457) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L134)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> We need to standardize on "self" or "that" for dealing with scope issues.</details>
I prefer `that`, because it is the preferred style in "Javascript: The Good Parts" (funny image: http://blog.movereem.nl/JavaScript-sources/ )

---

*[HTML5-frontend/ndt-library.js, line 136 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-136:-KHvH49I39VAHO0EV4_E:-625174913) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L136)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> Declarations at start of function.</details>
jslint will catch this!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/54#-:-KHvHn5LjoMkWEthTCTt:-514131530)*
<!-- Sent from Reviewable.io -->

pboothe commented:
*[HTML5-frontend/ndt-library.js, line 70 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-70:-KHvN93bui6bDjMz9-PV:896486616) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L70)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> already false in instantiation? </details>
can be shortened and clarified to 
```javascript
try {
  var ndt_js = new NDTjs();
  hasWebsockets = ndt_js.checkBrowserSupport();
} catch(e) {
  hasWebSockets = false;
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/54)*
<!-- Sent from Reviewable.io -->

shredtechular commented:
Ran JSLint on the ndt-library.js & script.js files and addressed the issues it found
![Screen Shot 2016-05-31 at 12.39.05 PM.png](https://files.reviewableusercontent.io/images/reviewable/2396774/FDF7xww_pc_BRETaxuteGMJY/Screen%20Shot%202016-05-31%20at%2012.39.05%20PM.png)

<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> Please put 
> `/*jslint bitwise: true, browser: true */` 
> after the introductory comment in every file (if it's not already there) then run the JSlint commandline tool (or use the jslint website, but the commandline too is easier to run multiple times).
</details>


---

Review status: 0 of 5 files reviewed at latest revision, 35 unresolved discussions, all commit checks successful.

---

*[HTML5-frontend/ndt-library.js, line 2 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-2:-KJ2k48POlZtc9pt7Z1l:1629176318) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L2)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> I'd like to call a moratorium on the word Interface here, where we could be using it in both the sense of GUI and in the independent sense of "the abstract object and function signatures which are fleshed out somewhere else".
</details>
The intention of this code is an object that has methods for a custom front end to utilize to run NDT tests.  So, I'm not really sure what I should do with these comments.  Are you suggesting something else for the file overview?

---

*[HTML5-frontend/ndt-library.js, line 19 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-19:-KJ2kZ6nkTzx0VhI6k0J:-1610106287) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L19)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> NDTJSInterface?
</details>
Not sure what this means, are you saying want to rename the object NDTJSInterface?  Above comment says not to use "Interface".  Do you want to call it something else?

---

*[HTML5-frontend/ndt-library.js, line 24 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-24:-KJ3R1cGbM9j59A_2Z8R:-1815537561) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L24)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> This is asynchronous, right? If so, this won't block and the rest might not be populated as it progresses.
</details>
That's correct, it is asynchronous.  We have error trapping if it is not ready and the front end can call the function again.

---

*[HTML5-frontend/ndt-library.js, line 28 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-28:-KJ2trmc7CR2S2AVtdaC:1036398519) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L28)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> "Private properties and methods should be named with a trailing underscore.
> Protected properties and methods should be named without a trailing underscore (like public ones)."
> 
> https://google.github.io/styleguide/javascriptguide.xml?showone=Code_formatting#Naming
> 
> Also, "We recommend the use of the JSDoc annotations @private and @protected to indicate visibility levels for classes, functions, and properties." 
> 
> https://google.github.io/styleguide/javascriptguide.xml?showone=Code_formatting#Visibility__private_and_protected_fields_
</details>
Added a trailing underscore to the private props and methods of the object and annotated them with @private, however (JSLint doesn't seem to like trailing underscores)[https://jslinterrors.com/unexpected-dangling-_-in-a].

---

*[HTML5-frontend/ndt-library.js, line 38 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-38:-KJ2u1M3a5_HKVUsr8BT:1740281843) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L38)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> Agree, please just delete the line.
</details>
Removed.

---

*[HTML5-frontend/ndt-library.js, line 40 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-40:-KJ3RO4jBGDpdhLsWNu0:1328548651) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L40)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> A better pattern to use here is to return the test object or to throw an exception if the browser doesn't support websocket NDT.
</details>
Added an exception and now returns an object.

---

*[HTML5-frontend/ndt-library.js, line 44 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-44:-KJ6dtg7RnzGCBDvP1K4:-1238664940) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L44)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> I agree here. Please don't use global variables unless you absolutely must.
</details>
changed to not using global here, instead object property

---

*[HTML5-frontend/ndt-library.js, line 46 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-46:-KJ6e4RKg9-S2r3TYCO-:1420478295) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L46)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> var declaraction
</details>
Became N/A after fixing other issues.

---

*[HTML5-frontend/ndt-library.js, line 50 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-50:-KJ3S3FVcEUOfg50hJDx:1185274928) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L50)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> Hopefully refactoring this to use exceptions will make it clearer.
</details>
refactored to use exceptions now.

---

*[HTML5-frontend/ndt-library.js, line 64 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-64:-KJ6e8HjX7NLJ4q-mjk8:-1483025915) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L64)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> sets to false, then true, then false, I would assume false and then find if true.
</details>
Fixed per @pboothe's suggestion

---

*[HTML5-frontend/ndt-library.js, line 70 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-70:-KJ3cEGtM4Q0m5DANZdU:1647499722) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L70)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> can be shortened and clarified to 
> ```javascript
> try {
>   var ndt_js = new NDTjs();
>   hasWebsockets = ndt_js.checkBrowserSupport();
> } catch(e) {
>   hasWebSockets = false;
> }
> ```
</details>
Fixed and returning true/false

---

*[HTML5-frontend/ndt-library.js, line 77 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-77:-KJ3cMZHHdXoKDwwR08V:-1420958076) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L77)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> if you assume true, then you might not need this line
</details>
Simplified function based on @pboothe's suggestion.

---

*[HTML5-frontend/ndt-library.js, line 81 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-81:-KJ3cdQsZ9u2YjidJei2:671757985) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L81)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> I think that offering up a Java Applet is confusing in the year 2016, and we should not do it unless we must.
</details>
Simplified function based on @pboothe's suggestion.  The whole point of this PR is to separate front end from backend in existing js.  I think removing functionality/tests should be a separate PR.

---

*[HTML5-frontend/ndt-library.js, line 82 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-82:-KJ3cnP_iVwFJqArviio:1758359711) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L82)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> Empty space.
</details>
Fixed.

---

*[HTML5-frontend/ndt-library.js, line 86 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-86:-KJ3cou5Z2d0ykgM9pu2:1758359711) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L86)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> empty space
</details>
Fixed.

---

*[HTML5-frontend/ndt-library.js, line 88 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-88:-KJ3cprWONo-6NM-NmaN:-1420958076) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L88)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> if neither hasWebsocket or hasJava?
</details>
Simplified function based on @pboothe's suggestion.

---

*[HTML5-frontend/ndt-library.js, line 93 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-93:-KJ3dDkUweqOFoce1oAs:1758359711) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L93)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> It's more of an HTML entity and not information, right?
</details>
Fixed.

---

*[HTML5-frontend/ndt-library.js, line 94 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-94:-KJ3dFKkiX2xwFrRRs4X:1758359711) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L94)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> Where does "false" come from – looks like True or the HTML entity?
</details>
Fixed.

---

*[HTML5-frontend/ndt-library.js, line 97 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-97:-KJ6eJBSaF-4QY7FRlHR:-1518033216) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L97)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> I would opt for these functions to be populated with variables like use_websocket_client passed as parameters, rather that use instance properties
</details>
Changed this instance to parameter and reviewed the rest to see if there were other instances like this.  I think in some of the other methods it makes sense to use instance properties to limit what an implementer would need to do, so I think this is better now?

---

*[HTML5-frontend/ndt-library.js, line 102 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-102:-KJ3lSsDh5RBlM-YV2ra:-497468049) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L102)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> var declaration beyond start of function.
</details>
Fixed

---

*[HTML5-frontend/ndt-library.js, line 117 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-117:-KJ3lYAHdfHIZB665Ltz:-497468049) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L117)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> "Object"?
</details>
Fixed

---

*[HTML5-frontend/ndt-library.js, line 122 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHvFpbitLTcvLILkrBF:-KJ3lfjMLnaDZ_cfwKP4:16681732) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L122)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> This is why we have `else {`. This is not a guard, instead the function is 
> ```javascript
> if (x) {
>   a();
> } else {
>   b();
> }```
> 
> When I describe this function in english, I find myself naturally using the word "otherwise", which means it needs an `else`.
</details>
JSLint doesn't like the else and returns "Unnecessary 'else' after disruption." if I try:

    ``if (this.websocket_client) {
           return this.websocket_client;
       } else {
           return $('#NDT');
       }``

Were you wanting something else here?

---

*[HTML5-frontend/ndt-library.js, line 132 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-132:-KJ3luIhG_-3Y89KGBG5:1197973799) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L132)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> Reads better with parens (because nobody remembers operator precedence rules outside of PEMDAS).
> ```javascript
> var mlabNsService = ('https:' == location.protocol) ? 'ndt_ssl' : 'ndt';
> ```
</details>
Fixed, used suggestion with parens.

---

*[HTML5-frontend/ndt-library.js, line 134 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-134:-KJ3lyqUGO9RL-mW6tnQ:1290184689) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L134)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> I prefer `that`, because it is the preferred style in "Javascript: The Good Parts" (funny image: http://blog.movereem.nl/JavaScript-sources/ )
</details>
OK, changed to `that`.

---

*[HTML5-frontend/ndt-library.js, line 136 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-136:-KJ6cePA0oDnGjncpVV6:1758359711) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L136)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> jslint will catch this!
</details>
Fixed.

---

*[HTML5-frontend/ndt-library.js, line 141 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-141:-KJ6cfhff3Fo06x_QEqp:1758359711) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L141)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> Declarations at start of function.
</details>
Fixed.

---

*[HTML5-frontend/ndt-library.js, line 159 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-159:-KJ3oAWyMkhvgugPBtxZ:1758359711) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L159)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> string
</details>
Fixed.

---

*[HTML5-frontend/ndt-library.js, line 167 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-167:-KJ3oBM4Sk2zgKoLCC36:1758359711) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L167)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> string
</details>
Fixed.

---

*[HTML5-frontend/ndt-library.js, line 188 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-188:-KJ3oQP4TmM_uzfK6IsV:1758359711) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L188)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> var declaration beyond start of function.
</details>
Fixed.

---

*[HTML5-frontend/ndt-library.js, line 231 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-231:-KJ3oqFNyTAf_tt0KY7N:-2034515506) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L231)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> As mentioned elsewhere, can this be done outside of a property of window?
</details>
I thought the point of this entire PR was to make it as easy as possible for an implementer to make use of an NDT "library".  From my perspective it seems like this would be easiest for an implementer rather than making them instantiate it?  And since this would be properly namespaced with all other methods/properties attached to this object, it seems like this is okay from the (google doc)[https://google.github.io/styleguide/javascriptguide.xml?showone=Naming#Naming]?  Maybe I'm not understanding what you guys have in mind here though?

"Use namespaces for global code
ALWAYS prefix identifiers in the global scope with a unique pseudo namespace related to the project or library."

---

*[HTML5-frontend/script.js, line 8 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8h-r1-8:-KJ3sQUv2h1vp-8x51QP:287594832) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/script.js#L8)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> Can we at this point change the script ("script.js") filename to something that is more understandable?
</details>
We can, but trying to limit the amount of changes for this PR and all this is designed to do is to separate frontend from backend code.  Plus, as I understand it the process says, 

"Don’t mix code moves/renames with other changes
Moving code around needs to be done in its own pull request because otherwise the reviewer can’t easily see what changed in the code after it was moved. The diff can’t distinguish between move and move+change."

https://docs.google.com/document/d/1Uhd_BUiwwNgBd4KcyfBKqUhOvuEZMAoqOtWCVXVbvkE/edit#

---

*[HTML5-frontend/script.js, line 65 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8h-r1-65:-KJ6dIqsLtRGkXRYnYnV:1758359711) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/script.js#L65)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> var declaration beyond start of function.
</details>
Fixed.

---

*[HTML5-frontend/script.js, line 452 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8h-r1-452:-KJ6dRv7NpKcJXwwlfU_:1758359711) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/script.js#L452)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> var declaration beyond start of function.
</details>
Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/54#-:-KJ6flSmGWH3egdcaizZ:-2011628855)*
<!-- Sent from Reviewable.io -->

collina commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[HTML5-frontend/ndt-library.js, line 2 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-2:-KJHpxnmiPm2RT4sAi7b:-580370643) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L2)):*
<details><summary><i>Previously, shredtechular (Andrew Newhouse) wrote…</i></summary>
> The intention of this code is an object that has methods for a custom front end to utilize to run NDT tests.  So, I'm not really sure what I should do with these comments.  Are you suggesting something else for the file overview?
</details>
While Peter disagrees with my use of Interface, my broader concern was the labeling as "library" – it's more an intermediary between NDTjs and the web page, so if you could find another name than library, I think it would be helpful.

---

*[HTML5-frontend/ndt-library.js, line 19 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-19:-KJHpfhnW_mcfpbjCIPW:1685513113) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L19)):*
<details><summary><i>Previously, shredtechular (Andrew Newhouse) wrote…</i></summary>
> Not sure what this means, are you saying want to rename the object NDTJSInterface?  Above comment says not to use "Interface".  Do you want to call it something else?
</details>
Yes, something else.

---

*[HTML5-frontend/ndt-library.js, line 81 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-81:-KJHr76iQ6pKRHJbXGL_:-1530771663) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L81)):*
<details><summary><i>Previously, shredtechular (Andrew Newhouse) wrote…</i></summary>
> Simplified function based on @pboothe's suggestion.  The whole point of this PR is to separate front end from backend in existing js.  I think removing functionality/tests should be a separate PR.
</details>
Agreed.

---

*[HTML5-frontend/ndt-library.js, line 122 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHvFpbitLTcvLILkrBF:-KJHrSzO9gET-G5gn-0R:-203771709) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L122)):*
<details><summary><i>Previously, shredtechular (Andrew Newhouse) wrote…</i></summary>
> JSLint doesn't like the else and returns "Unnecessary 'else' after disruption." if I try:
> 
>     ``if (this.websocket_client) {
>            return this.websocket_client;
>        } else {
>            return $('#NDT');
>        }``
> 
> Were you wanting something else here?
</details>
JSLint seems correct here, I think the original code was correct.

---

*[HTML5-frontend/ndt-library.js, line 231 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-231:-KJHriChs-oYQMY-j6Ll:1717516509) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L231)):*
<details><summary><i>Previously, shredtechular (Andrew Newhouse) wrote…</i></summary>
> I thought the point of this entire PR was to make it as easy as possible for an implementer to make use of an NDT "library".  From my perspective it seems like this would be easiest for an implementer rather than making them instantiate it?  And since this would be properly namespaced with all other methods/properties attached to this object, it seems like this is okay from the (google doc)[https://google.github.io/styleguide/javascriptguide.xml?showone=Naming#Naming]?  Maybe I'm not understanding what you guys have in mind here though?
> 
> "Use namespaces for global code
> ALWAYS prefix identifiers in the global scope with a unique pseudo namespace related to the project or library."
</details>
We can punt this to later.

---

*[HTML5-frontend/script.js, line 8 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8h-r1-8:-KJHrnMQez8mpre1qq2D:682910885) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/script.js#L8)):*
<details><summary><i>Previously, shredtechular (Andrew Newhouse) wrote…</i></summary>
> We can, but trying to limit the amount of changes for this PR and all this is designed to do is to separate frontend from backend code.  Plus, as I understand it the process says, 
> 
> "Don’t mix code moves/renames with other changes
> Moving code around needs to be done in its own pull request because otherwise the reviewer can’t easily see what changed in the code after it was moved. The diff can’t distinguish between move and move+change."
> 
> https://docs.google.com/document/d/1Uhd_BUiwwNgBd4KcyfBKqUhOvuEZMAoqOtWCVXVbvkE/edit#
</details>
Ack. If you can put in a PR after pull with the rename, I'll approve it right away.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/54)*
<!-- Sent from Reviewable.io -->

shredtechular commented:
Reviewed 1 of 5 files at r1.
Review status: 1 of 5 files reviewed at latest revision, 33 unresolved discussions, all commit checks successful.

---

*[HTML5-frontend/ndt-library.js, line 2 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-2:-KJpc6QyhN43FI5e4I-5:-42294240) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L2)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> While Peter disagrees with my use of Interface, my broader concern was the labeling as "library" – it's more an intermediary between NDTjs and the web page, so if you could find another name than library, I think it would be helpful.
</details>
Do you have a suggestion here if you don't like the "library" label?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/54)*
<!-- Sent from Reviewable.io -->

shredtechular commented:
Review status: 1 of 5 files reviewed at latest revision, 33 unresolved discussions, all commit checks successful.

---

*[HTML5-frontend/ndt-library.js, line 19 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/54#-KHkQGHDqbQM4SLs_U8k-r1-19:-KJpd3pjdLy4IzwB5Gtt:149579037) ([raw file](https://github.com/m-lab/ndt/blob/81ab6032f062529bcf360801bc800bca95ca9f11/HTML5-frontend/ndt-library.js#L19)):*
<details><summary><i>Previously, collina (Collin Anderson) wrote…</i></summary>
> Yes, something else.
</details>
If not Library or Interface, any other suggestions then?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/54)*
<!-- Sent from Reviewable.io -->
